### ISSUE TITLE
Enhance #set: parser function with display option

### ISSUE THREAD

**Issue by krabina** [SemanticMediaWiki/SemanticMediaWiki#5624]
Since the `#set:` parser function is not a 100% replacement for the inline annotation` [[Foo::Bar]]`, we should need to move forward in this direction. Therefore I propose to add a display functionality to the parser function.

### Background
In this issue, I am not referring to the problem of SMW installations that generally use the inline annotation to annotate content in wikitext using the wikitext editor. For this, see [#4727](https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/4727)

In the use case I am addressing, semantic values are set in a template, which is the general use case for many SMW installations. It is generally not a problem to avoid inline annoations in templates: 
Instead of putting 
`Your Property: [[Foo::Bar]]` in your template (maybe for a fact box of a wiki article), it is easy to do something like
`Your Property: {{{Bar|}}}{{#set:Foo={{{Bar|}}} }}`
This approach also has the advantage, that in your template you can control the output of the value `{{{Bar|}}}` anyway you like and set the property with the silent `{{#set:Foo={{{Bar|}}} }}` annotation.

### Problem
The problem, however, is that in the set parser function is not a complete replacement for the inline annotation. These are some cases (there are certainly more):
- Datatype External Identifier will render a link that points to the external formatter URI. You cannot achieve this with `[[{{{Bar|}}}]] `in the template. This will render an internal link instead of the link to the external URI.
- Datatype Keyword can be used with a [keyword link formatter rule](https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/3019), which also works only with inline annotations.
- The [property link feature](https://github.com/SemanticMediaWiki/SemanticMediaWiki/issues/1855) only works with inline annotations: `[[Your Property::@@@]]`
- It is cumbersome to annotate Property Pages with the silent annotation, since you usually want have something displayed there. My workaround is this, e. g. for a page Property:Link

```
* [[Has type::@@@]]: {{#show:{{FULLPAGENAME}}|?Has type}}
* [[Has preferred property label::@@@]]]: {{#show:{{FULLPAGENAME}}|?Has preferred property label}}
* [[Has property description::@@@]]: {{#show:{{FULLPAGENAME}}|?Has property description}}

{{#set:
|Has type=Url
|Has preferred property label=Link@de
|Has property description=Externer Link@de
|Has property description=External Link@en
}}
```
### Solution
It should not be very hard to add a display option to the #set parser function. Basically it is something like combing #set and #show. 
Suggested syntax:
```
{{#set:
|Has type=Url|+display=link
}}
```
Other options:
- none (default)
- text 

This approach would also enable backwards-compatibility, because if the option is omitted it can default to "+display=no" which does not change the current behavior.

I am interested to hear what others think about this.

**Comment by krabina** [comment]
@cscott  - I just rewatched your presentation at MWCon. This issue mentions some places where the #set parser function is not yet a replacement for the inline annotations

**Comment by YOUR1** [comment]
The first question coming up to mind is how does this work with a LUA combo?

**Comment by gesinn-it-gea** [comment]
Why not using https://www.semantic-mediawiki.org/wiki/Help:Setting_values/Working_with_the_template_parameter

`{{#set:Comment|Alexander|template=Text}}`

and a template with

`{{{value}}}`

**Comment by krabina** [comment]
It might be a disadvantage for performance if you always have to use template inclusion in every property annotation.

**Comment by Sheep-y** [comment]
I want the unit conversion hint provided by the annotation syntax, which is not passed to the template.  Or the link of page properties, which would require a new template.

So I took a look, the code change is relatively simple.  In `src\ParserFunctions\SetParserFunction.php`->`parse(...)`, instead of calling $templateRenderer, change to append `$dataValue->getShortWikitext()`.

[Attached](https://github.com/user-attachments/files/16371830/SMW.5624.by.Sheepy.zip) my diff and the modded file. Error messages is retained, and the glue can be specified for multiple values.

* `{{#set: A=Foo | template=# }}` renders as `Foo` (exact format depends on property type)
* `{{#set: | A=Foo | B=Bar | template=# }}` renders as `FooBar` (ditto)
* `{{#set: | A=Foo,Bar | +sep=, | template=# }}` also renders as `FooBar` (ditto)
* `{{#set: | A=Foo | B=Bar | A=Baz | template=#,&#32 }}` renders as `Foo, Baz, Bar` (ditto)

Specification:
* Pass `#` to `template` to render property values.
* Add text after hash to insert between values.
* Values are displayed in order of property declaration, then in order of value declaration (same as using a real template).

Advantage:
* No new property conflicts.
* * e.g. if a wiki has Property:display, it cannot be #set had we added a display parameter.
* No template building and parsing.
* Compatible with future MediaWiki (Parsoid).

Disadvantage:
* Cannot set value display like annotation.
* Break templates that starts with `#`, but only for `#set`.
* The template parameter's value is trimmed, so `&#32;` has to be used for the glue to ends with space.
* The # syntax is inspired by #ask's Plain (unformatted) printouts.  Which may cause confusion as the result here is formatted.  For unformatted result, use a `{{{value}}}` template.

**Comment by JeroenDeDauw** [comment]
> Property link `[[Property::@@@]]` is another problem unrelated to setting property values, and IMHO it should be solved with a new function or tag.

Agreed. Something like `{{#property_link:Has type}}` is cleaner than the current approach of having a magical special value in the standard link syntax.

**Comment by JeroenDeDauw** [comment]
I think using a dedicated new parameter for specifying what `#set` should output is better than introducing a new case with its own syntax to the `template` parameter.

How about this?

* `{{ #set: A=Foo | output=$ }}` outputs `Foo`
* `{{ #set: A=Foo | output=Your Hardcoded Text }}` equivalent to [[A::Foo|Your Hardcoded Text]]
* `{{ #set: A=Foo | B=Bar | output=$ }}` outputs `FooBar`
* `{{ #set: A=Foo | B=Bar | output="$", }}` outputs `"Foo", "Bar",`

That would be a breaking change because people might have a property called `output`, or whatever we decide to call that parameter.

Another approach is a new dedicated function for setting and outputting a value. 

* `{{ #set_and_show: A=Foo }}` outputs `Foo`
* `{{ #set_and_show: A=Foo | output=Your Hardcoded Text }}` equivalent to [[A::Foo|Your Hardcoded Text]]
* `{{ #set_and_show: A=Foo | B=Bar }}` outputs `FooBar`, and so does `{{ #set_and_show: A=Foo | B=Bar | output=$ }}` 
* `{{ #set_and_show: A=Foo | B=Bar | output="$", }}` outputs `"Foo", "Bar",` (could _also_ add a separator parameter)

**Comment by krabina** [comment]
how about `setv` for "set verbose" in contrast to silent. or "set value". 
But is there not also a big disadvantage? With `set and show` you have to implement (and maintain) everything that set does. With an additional parameter to set there is only the new functionality to add.

and my suggested syntax
`{{#set:
|Has type=Url|+display=
}}`
would not be a breakting change since you can have the variable display set to anything and +display indicating the display options.
